api/envoy/extensions/filters/http/grpc_field_extraction/v3/config.proto

```diff
@@ -0,0 +1,143 @@
+//
+// # Assumptions
+// 1. this filter is only applicable for gRPC with Protobuf as payload.
+// 2. for bi-directional and client-side gRPC streaming, the initial message from the client should not depend on receiving server initial metadata.
+//
+// # Process Flow
+// When a request reaches the filter, it will check
+// 1. if the request is a gRPC request configured for extraction, the filter tries:
+//    a. block headers and buffer the incoming decode before getting the first complete gRPC message
+//    b. look up the target field from the buffered gRPC message
+//    c. if the extraction result isn't empty, write it into the dynamic metadata and resume the request propagation.
+// 2. otherwise, pass through the request.
+//
+// If the request is a malformed one found during 1.a or 1.b, the filter will reject the request.
+//
+// # Config Requirements
+// 1. the target field should be singular primitive type or repeated primitive
+// type and its value will be extracted in string format.
+// 2. the intermediate type could also be repeated.
+//
+// # Output Format
+// The result format will be `field<StringValue>` -> `values<ListValue of StringValue>` in the dynamic metadata<google.protobuf.Struct>.
+//
+// # Performance
+// This filter should be performant as it
+// 1. converts between the gRPC message from EnvoyBuffer without data copy.
+// 2. parse the gRPC message binary directly without deserialization.
+//
+// # Example,
+// we have the following request definition for the gRPC method `/pkg.svc/Method`.
+//
+// message MethodRequest {
+//   string foo = 1;
+//   Nested nested = 2;
+//   ...
+// }
+//
+// message Nested {
+//  repeated string bar = 1;
+// }
+//
+// This is the filter config(expressed in JSON format).
+// {
+//   "descriptor_set":{...},
+//   "extractions_by_method": {
+//      "/pkg.svc/Method":{
+//         "request_field_extractions":{
+//            "foo":{
+//            },
+//            "nested.bar":{
+//            }
+//         }
+//      }
+//   },
+//   ...
+// }
+//
+//
+// During runtime, the filter receives the following `MethodRequest` message(expressed in JSON format).
+// {
+//   foo: "val_foo",
+//   nested: { "bar": ["val_bar1", "val_bar2"]}
+// }
+//
+// The filter will writes the following dynamic metadata(expressed in JSON format).
+//
+// `envoy.filters.http.grpc_field_extraction`: {
+//   "foo":[
+//      "val_foo"
+//   ],
+//   "nested.bar":[
+//      "val_bar1", "val_bar2"
+//   ]
+// }
+
+message GrpcFieldExtractionConfig {
+  // The proto descriptor set binary for the gRPC services.
+  //
+  // It could be passed by a local file through `Datasource.filename` or embedded in the
+  // `Datasource.inline_bytes`.
+  config.core.v3.DataSource descriptor_set = 1 [(validate.rules).message = {required: true}];
+
+  // Specify the extraction info.
+  // The key is the fully qualified gRPC method name.
+  // `/$package.$Service/$Method`, like
```

Comment:
I thought we would use selector syntax instead of gRPC method name? Why switch to gRPC method name?

Reply:
In fact, grpc transcoder also uses selector-style syntax for service name

https://github.com/envoyproxy/envoy/blob/701aa7858e78b3bcad7f1fb1ac5ea880603c92a2/api/envoy/extensions/filters/http/grpc_json_transcoder/v3/transcoder.proto#L127

example: `google.example.apikeys.v1.ApiKeys`

Reply:
Updated. 

Maybe better call it protobuf method name style as selector isn't known outside google3. Originally by https://github.com/grpc/grpc-java/issues/7651 I see gRPC method style is more recommended but I guess either one should be fine.

Reply:
I think expressing this as a path string will actually be more efficient, because the data plane can just do a string match.

If you want to be more flexible here, an alternative would be to use the [generic matching API](https://github.com/cncf/xds/blob/e9ce68804cb4e64cab5a52e3c8baf840d4ff87b7/xds/type/matcher/v3/matcher.proto#L23), so that the filter can be configured to match on arbitrary criteria of the request.

Reply:
True it is more efficient, but during filter init we can just replace the `.` with `/` and store the formatted string for matching.

For the API, we should focus on standardization for user experience, not performance. With path-based config, there are some nits - what happens if user adds a trailing slash for example?

Reply:
I don't see how that's any different from any other path-matching semantic.  If you add a trailing slash and the path doesn't have one, it won't match.

The path semantic seems much clearer to me.  But I'll leave it up to you.

Reply:
Thanks Mark and Teju!

Unfortunately using gRPC path/protobuf path won't make difference as proto descriptor use protobuf path so eventually we need to have one.